web/src/hooks/queries/useHomePageBlockQuery.ts

```diff
@@ -0,0 +1,138 @@
+            ...c,
+            numberDisputes: c.numberDisputes,
+            treeNumberDisputes: c.numberDisputes,
+            numberVotes: c.numberVotes,
+            treeNumberVotes: c.numberVotes,
+            stake: c.stake,
+          }))
+        : usedQuery.data.pastCourts.map((c, i) => ({
+            ...c,
+            numberDisputes: usedQuery.data.presentCourts[i].numberDisputes - c.numberDisputes,
+            treeNumberDisputes: usedQuery.data.presentCourts[i].numberDisputes - c.numberDisputes,
+            numberVotes: usedQuery.data.presentCourts[i].numberVotes - c.numberVotes,
+            treeNumberVotes: usedQuery.data.presentCourts[i].numberVotes - c.numberVotes,
+            stake: (BigInt(usedQuery.data.presentCourts[i].stake) + BigInt(c.stake)) / 2n,
+          }));
+      const mostDisputedCourt = diffCourts.sort((a, b) => b.numberDisputes - a.numberDisputes)[0];
+      // 2. biggest chances of getting drawn
+      // fact a: getting drawn in a parent court also subjects you to its rewards
+      // fact b: staking in children, stakes in parents. but subgraph at this date doesn't reflect this
+      // so, stakes trickle up, rewards/disputes trickle down
+
+      for (const parent of diffCourts) {
+        for (const child of diffCourts) {
+          if (parent.id === child.parent?.id) {
+            child.treeNumberVotes = String(Number(parent.treeNumberVotes) + Number(child.treeNumberVotes));
+          }
+        }
+      }
+      diffCourts.reverse();
+      for (const child of diffCourts) {
+        for (const parent of diffCourts) {
+          if (parent.id === child.parent?.id) {
+            parent.stake = String(BigInt(parent.stake) + BigInt(child.stake));
+          }
+        }
+      }
+      diffCourts.reverse();
+      for (const c of diffCourts) {
+        c.votesPerPnk = Number(c.numberVotes) / (Number(c.stake) / 1e18);
+        c.treeVotesPerPnk = c.votesPerPnk;
+      }
+      for (const parent of diffCourts) {
+        for (const child of diffCourts) {
+          if (parent.id === child.parent?.id) {
+            child.treeVotesPerPnk += parent.votesPerPnk;
+          }
+        }
+      }
+      const bestDrawingChancesCourt = diffCourts.sort((a, b) => b.treeVotesPerPnk - a.treeVotesPerPnk)[0];
+      // 3. expected reward
+      // since we isolated the exclusive disputes from the cumulative disputes
+      // we can calculate the "isolated reward" of every court
+      // after that's done, then just trickle the rewards down
+
+      for (const c of diffCourts) {
+        c.expectedRewardPerPnk = c.votesPerPnk * c.feeForJuror;
+        c.treeExpectedRewardPerPnk = c.expectedRewardPerPnk;
+      }
+      for (const parent of diffCourts) {
+        for (const child of diffCourts) {
+          if (parent.id === child.parent?.id) {
+            child.treeExpectedRewardPerPnk = parent.treeExpectedRewardPerPnk + child.treeExpectedRewardPerPnk;
+          }
+        }
+      }
```

Comment:
Here we are abusing loops, I would try to optimize them with something like this (pseudo code):
```typescript
const processed = new Set();

const process = (court) => {
  if (!processed.has(court.id) && court.parent) {
    court.treeNumberVotes = court.treeNumberVotes + process(court.parent);
  }
  processed.add(court.id);
  return court.treeNumberVotes;
}

for (const court of diffCourts.toReversed()) {
  if (!processed.has(court.id))
    process(court);
}
```
I would need to test but with this I think we can do it in linear time, also we might be able to do all the calculations with only one loop.